### ISSUE TITLE
Bernoulli(logits=-float('inf')).log_prob(torch.FloatTensor([1])) returns nan

### ISSUE THREAD

**Issue by alicanb** [probtorch/pytorch#82]
@neerajprad can you take a look?

**Comment by neerajprad** [comment]
@alicanb - This is because logits are used as is without clamping. This was based on the assumption that providing 0 as prob values from the user might be common, but when the user provides logits they are probably directly using the output of a neural net, and we can offload the onus of doing inf checks to them. That's the reason why the tests too use very high/low values for logits but not `infs`. Ideally, I agree that the behavior should be exactly the same. For this, I can add some validation check in the constructor, and provide minimum value from Finfo to be used instead. These checks can get expensive though, and ideally we should also support a mode without any clamping. What do you think?

**Comment by alicanb** [comment]
We should think about what we want to return `log(0)`: `log(eps)` or `-inf`? If we are ok with log(eps) then let's clamp logits to [log(eps), log(1-eps)] like we do do probs. This would get rid of the mutating issue in multinomial.

If we support a mode without clamping, I don't know how to make things consistent. How about let's clamp logits as well for now, and write some cases in the design doc how we want the non-clamping mode to behave? I'm assuming by then multinomial and binomial will be merged so we can change behaviour in all distributions that support logits at once.
